File: datasets/C++/Chatgpt/DataStructures/medium5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val): data(val), left(nullptr), right(nullptr) {}
};

bool isBST(Node* node, int minVal, int maxVal) {
    if (!node) return true;
    if (node->data <= minVal || node->data >= maxVal)
        return false;
    return isBST(node->left, minVal, node->data) &&
           isBST(node->right, node->data, maxVal);
}

Node* buildTree(const vector<int>& arr) {
    if (arr.empty() || arr[0] == -1) return nullptr;
    Node* root = new Node(arr[0]);
    queue<Node*> q;
    q.push(root);
    int i = 1;

    while (!q.empty() && i < arr.size()) {
        Node* curr = q.front(); q.pop();

        if (i < arr.size() && arr[i] != -1) {
            curr->left = new Node(arr[i]);
            q.push(curr->left);
        }
        i++;
        if (i < arr.size() && arr[i] != -1) {
            curr->right = new Node(arr[i]);
            q.push(curr->right);
        }
        i++;
    }
    return root;
}

int main() {
    cout << "Enter nodes in level order (-1 for null): ";
    vector<int> arr;
    int val;
    while (cin.peek() != '\n' && cin >> val) {
        arr.push_back(val);
    }
    Node* root = buildTree(arr);
    cout << (isBST(root, INT_MIN, INT_MAX) ? "Yes" : "No") << endl;
    return 0;
}

File: datasets/Python/Chatgpt/DataStructures/medium5.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def isBST(node, min_val=float('-inf'), max_val=float('inf')):
    if node is None:
        return True
    if not (min_val < node.data < max_val):
        return False
    return (isBST(node.left, min_val, node.data) and 
            isBST(node.right, node.data, max_val))

def buildTree():
    from collections import deque
    print("Enter the binary tree nodes in level order (use -1 for null):")
    arr = list(map(int, input().split()))
    if not arr or arr[0] == -1:
        return None

    root = Node(arr[0])
    q = deque([root])
    idx = 1
    while q and idx < len(arr):
        current = q.popleft()

        if idx < len(arr) and arr[idx] != -1:
            current.left = Node(arr[idx])
            q.append(current.left)
        idx += 1

        if idx < len(arr) and arr[idx] != -1:
            current.right = Node(arr[idx])
            q.append(current.right)
        idx += 1

    return root

if __name__ == "__main__":
    root = buildTree()
    print("Yes" if isBST(root) else "No")
